block-storage/src/main/scala/coop/rchain/blockstorage/dag/Finalizer.scala

```diff
@@ -0,0 +1,183 @@
+package coop.rchain.blockstorage.dag
+
+import cats.syntax.all._
+
+import scala.collection.compat.immutable.LazyList
+
+/**
+  * Message (view) with all necessary data for finalizer
+  *
+  * @param id message unique identifier
+  * @param height message height (number)
+  * @param sender message sender identifier
+  * @param senderSeq sender's sequence number
+  * @param bondsMap bonds map used to calculate next fringe
+  * @param parents justification for the message
+  * @param fringe fringe observed by the message
+  * @param seen all previous messages seen by the message
+  * @tparam M abstract type for identifier of the message
+  * @tparam S abstract type for identifier of the sender
+  */
+final case class Message[M, S](
+    id: M,
+    height: Long,
+    sender: S,
+    senderSeq: Long,
+    bondsMap: Map[S, Long],
+    parents: Set[M],
+    fringe: Set[M],
+    // Cache of seen message ids
+    seen: Set[M]
+) {
+  override def hashCode(): Int = this.id.hashCode()
+}
+
+/**
+  * Multi-parent finalization implementation
+  *
+  * @param msgViewMap cache of all messages accessed by message id
+  * @tparam M abstract type for identifier of the message
+  * @tparam S abstract type for identifier of the sender
+  */
+final case class Finalizer[M, S](msgViewMap: Map[M, Message[M, S]]) {
+  // Helper function to lazily unfold iterator function
+  def unfold[A](init: A)(f: A => Iterator[A]): LazyList[A] =
+    LazyList.unfold(f(init)) { iter =>
+      if (!iter.hasNext) none
+      else {
+        val x = iter.next()
+        Some((x, iter ++ f(x)))
+      }
+    }
+
+  // Iterate self parent messages
+  def selfParents(mv: Message[M, S], finalized: Set[Message[M, S]]): Seq[Message[M, S]] =
```

Comment:
Just to make sure I understand correctly - this produces self parents only inside conflict set? So this finalized set along with filtering in `unfold` fuction works as a traversal stopper?

Reply:
Yes, all calculation for the next fringe should never go below existing fringe. As opposite what we have now where we need to download approved block justifications + self parent for each of them (which I also plan to remove).
https://github.com/rchain/rchain/blob/ecddeba9694184da11acf6cf4c7f1fba60ee6d86/casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala#L78-L89